### PR TITLE
[Orders with Coupons] Fix icon tooltip rendering when order has no discounts

### PR DESCRIPTION
**Pull request by iamgabrielma**
https://github.com/woocommerce/woocommerce-ios/pull/10956 should be merged first

## Description
This PR addresses a bug when rendering the tooltip icon on product discounts, which should appear only when an Order has coupons applied to it (do product discounts are disallowed and we show the informational tooltip to offer more info about it):

Screenshot 2023-10-17 at 18 03 24

## Changes
* The fix is only the line here: https://github.com/woocommerce/woocommerce-ios/commit/cafdf252fe0658824dda5f3321960735e1c953b6 . We check if discounts are disallowed, and if that's the case then we render the tooltip icon.
* Additionally we do some clean-up by:
  * We declare a new property `shouldDisallowDiscounts` that acts as a wrapper for `orderHasCoupons'. This encapsulates and hides the implementation details, and clarifies the logic. 
  * Add unit tests to both `EditableOrderViewModelTests` and `ProductRowViewModelTests`.
  * Extracting `TooltipView` to its own file.

## Testing instructions
* Go to Orders > + > `+ Add Product` > Select a product > Tap on the chevron icon to expand the product details.
* Observe that the `+ Add discount` row appears without informational tooltip
* Tap on `+ Add Coupon` and apply a valid coupon to the order
* Observe that the `+ Add discount` row is disabled and an informational tooltip can be seen and tapped.

| No discounts or coupons | Coupon applied |
|--------|--------|
| _Case 1 - No discounts or coupons - Icon hidden | _Case 3 - No discounts but a coupons - Icon appear |

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift [hunk 1]
```diff
@@ -403,7 +403,7 @@ private struct ProductsSection: View {
                 ForEach(viewModel.productRows) { productRow in
                     CollapsibleProductRowCard(viewModel: productRow,
                                               shouldDisableDiscountEditing: viewModel.paymentDataViewModel.isLoading,
-                                              shouldDisallowDiscounts: viewModel.orderHasCoupons,
+                                              shouldDisallowDiscounts: viewModel.shouldDisallowDiscounts,
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:406]: Ideally we'll move these properties to their respective view models, so the `CollapsibleProductRowCard` only needs the references to the order and to the product row to be initialized without having to care of the details, but this requires more extensive changes and testing so we've left it outside this PR.

**Comment by hafizrahman** [comment]
This works well in my test, I'm posting a couple of observations not related to this PR as separate messages below so they can easily be converted to separate issues, if needed.

**Comment by hafizrahman** [comment]
The tooltip message does not stick to the question mark icon, but instead shown under the discounted price:

**Comment by hafizrahman** [comment]
When removing a coupon, in order to be able to add a specific discount instead of coupon, the loading bar to the right is green. Not sure if intentional. But since the question mark icon its replacing is not green, I thought I'd mention it:

https://github.com/woocommerce/woocommerce-ios/assets/266376/22973392-14b8-4ae4-886c-af089e8817ee

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift (hunk 1 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:406]: Thanks for mentioning this, I have never worked with this screen before and it's interesting to see the approach of having separate viewmodels to handle the complexity of the data it's working with.

**Review by hafizrahman — APPROVED** [review]
Nice work with the fix, good thinking with refactoring the tooltip as well. 🚀

**Comment by iamgabrielma** [comment]
Thanks for the review!

> The tooltip message does not stick to the question mark icon, but instead shown under the discounted price

Oh! Thanks for pointing that out, this one went under the radar. Making the ToolTip adaptable to its parent view has been more challenging that anticipated when this view might render additional subviews, since we also need to modify the tooltip offset to account for it, there's a few alternatives that we can use once we move up from iOS15, so we don't have to create this view manually and deal with these edge cases: https://github.com/woocommerce/woocommerce-ios/issues/10917

> When removing a coupon, in order to be able to add a specific discount instead of coupon, the loading bar to the right is green. Not sure if intentional. But since the question mark icon its replacing is not green

Good catch! This uses the [.redacted(reason:) property](https://developer.apple.com/documentation/SwiftUI/View/redacted(reason:)) which we use to [disallow editing while the order synchronises remotely](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/CollapsibleProductRowCard.swift#L204-L206), this way we avoid that a merchant edits the field while we have not synced yet, as well as showing an out-of-date discount. We don't have much control over its styling since uses the underlying view where it's applied to (in this case the previous state, a green-styled discount, rather than the next state), but I'll take a further look to see if we can make it generic 👍

**Pull request merged**